src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -252,82 +254,133 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
 
     entry->CanonicalName = NULL;
     entry->Aliases = NULL;
-    entry->AddressListHandle = info;
-    entry->IPAddressCount = 0;    
+    entry->AddressList = NULL;
+    entry->AddressCount = 0;
 
     // Find the canonical name for this host (if any) and count the number of IP end points.
     for (struct addrinfo* ai = info; ai != NULL; ai = ai->ai_next)
     {
         // If we haven't found a canonical name yet and this addrinfo has one, copy it
         if ((entry->CanonicalName == NULL) && (ai->ai_canonname != NULL))
         {
-            entry->CanonicalName = (uint8_t*)ai->ai_canonname;
+            entry->CanonicalName = (uint8_t*)strdup(ai->ai_canonname);
+            if (entry->CanonicalName == NULL)
+            {
+                ConvertGetAddrInfoAndGetNameInfoErrorsToPal(EAI_MEMORY);
+            }
         }
 
         if (ai->ai_family == AF_INET || ai->ai_family == AF_INET6)
         {
-            entry->IPAddressCount++;
+            entry->AddressCount++;
         }
     }
 
-    return GetAddrInfoErrorFlags_EAI_SUCCESS;
-}
+#if HAVE_GETIFADDRS
+    char name[MAX_HOST_NAME + 1];
+    result = gethostname((char*)name, MAX_HOST_NAME);
```

Comment:
We will call this for every lookup, right? e.g. we will slow down common case when trying to resolve names other than ours. Since it is unlikely this will change during process run it may be worth of some caching.

Reply:
The same behavior, without caching, is on Mono.